### DISCUSSION TITLE
How to prevent brew install from automatically linking contents from a folder?

### DISCUSSION THREAD

**Discussion by cesarcoatl** [Homebrew/discussions#553] [Tap maintenance and Homebrew development]
Hello.

I have created formulae (not yet published), and I am going the "versioned" route, e.g. `foo@1.0.0`, `foo@2.0.0`, with the intention of installing and running both on the same machine.

But the problem I am facing is that some files and folders are linked at `/usr/local/lib`, so when I try to install the other version I get a warning and it is instructing me to execute `brew link --overwrite`, which would "unlink" the other version.

Is there something I can do on the `install` block or  anywhere else so both versions can coexist?

Thanks!

**Comment by carlocab** [comment]
Try using `keg_only :versioned_formula`. This will avoid symlinking them into `/usr/local`, but you can find them in `/usr/local/opt/foo@version`.

**Reply by cesarcoatl** [reply]
I see. But that prevents the creation of symlinks under `/usr/local/bin`, something that I am also intending here (see #546).

After installing both formulae I do see each one has its symlink under `/usr/local/opt`, and I do get this message:

```bash
$ brew install foo@1.0
...
==> Caveats
foo@1.0 is keg-only, which means it was not symlinked into /usr/local,
because this is an alternate version of another formula.

If you need to have foo@1.0 first in your PATH run:
  echo 'export PATH="/usr/local/opt/foo@1.0/bin:$PATH"' >> ~/.zshrc

For compilers to find foo@1.0 you may need to set:
  export LDFLAGS="-L/usr/local/opt/foo@1.0/lib"
```

I have no issue with updating my references from `/usr/local/bin` to `/usr/local/opt/foo@1.0/bin`, but just to make sure I clearly undersand. Is this the only option for this type of formulae?

**Reply by carlocab** [reply]
Another option is to manually rename things in `$(brew --cellar foo@version)/{bin,include,lib}` so that it doesn't conflict with what's there for `foo@another_version`.

You can have a look at the `lua@5.1` formula for an example of this.

**Reply by cesarcoatl** [reply]
This is basically what I've got for this formula:

```ruby
class FooAT10 < Formula
  desc "Awesome description"
  homepage "https://foo.com/"
  url "https://files.foo.com/release/foo/1.0/foo-1.0.zip", using: :curl, referer: "https://foo.com/"
  sha256 "7122870ff380ac0ffc22a81fd13ff7bc5de1a0736ac39f7bed9456cfec4b799a"
  license :cannot_represent

  def install
    prefix.install Dir["*"]
    # Make files executable and add symlinks
    %w[foo.sh bar.sh baz].each do |cmd|
      chmod "u=wrx,go=rx", "#{prefix}/#{cmd}"
      bin.install_symlink "#{prefix}/#{cmd}" => "#{cmd.gsub(".sh", "")}-1.0"
    end
    # Update plist file
    plist = "#{prefix}/com.foo.plist"
    text = File.read(plist)
    replace = text.gsub("/usr/local/bin/foo", "/usr/local/bin/foo-1.0")
    File.open(plist, "w") { |file| file.puts replace }
  end

  test do
    assert_match "Foo", shell_output("#{bin}/foo-1.0 status")
  end
end
```

**Comment by cesarcoatl** [comment]
Additional info.

The ZIP file, once unpacked, contains a `lib` directory and the install process automatically symlinks all files and directories.

Is there a way to prevent this?

```bash
$ brew install --debug --verbose foo@1.0
...
==> Finishing up
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/my-tap/homebrew-my-tap/Formula/foo@2.0.rb
ln -s ../Cellar/foo@1.0/1.0/bin/foo-1.0 foo-1.0
ln -s ../Cellar/foo@1.0/1.0/bin/bar-1.0 bar-1.0
ln -s ../Cellar/foo@1.0/1.0/bin/baz-1.0 baz-1.0
ln -s ../Cellar/foo@1.0/1.0/lib/core core
ln -s ../Cellar/foo@1.0/1.0/lib/install-info.txt install-info.txt
ln -s ../Cellar/foo@1.0/1.0/lib/libfoo.dylib libfoo.dylib
ln -s ../Cellar/foo@1.0/1.0/lib/libwrapper-macosx-universal-64.dylib libwrapper-macosx-universal-64.dylib
ln -s ../Cellar/foo@1.0/1.0/lib/libwrapper-macosx-universal-64.jnilib libwrapper-macosx-universal-64.jnilib
ln -s ../Cellar/foo@1.0/1.0/lib/opt opt
ln -s ../Cellar/foo@1.0/1.0/lib/runtime runtime
ln -s ../Cellar/foo@1.0/1.0/lib/wrapper.jar wrapper.jar
...
```

Additionally. Why is there a mention of `foo@2.0.rb`?

Thanks

**Reply by cesarcoatl** [reply]
Changed the title because of this finding.

**Reply by jonchang** [reply]
You might find it easier to just install everything into libexec and manually linking what you want into the appropriate directories

```ruby
def install
  libexec.install Dir[*]
  lib.install_symlink "#{libexec}/foo.dylib"
  # ...
end
```

**Reply by cesarcoatl** [reply]
Awesome!

This got me 99.999% of the way. I only have one final thing to ask.

There are some files that are deflated onto `prefix` that I wish to move to `libexec`.

I had this on my Formula as a final step in the `install` block, but it produces an error:

```ruby
    mv "#{prefix}/License.html", "#{libexec}/License.html"
    mv "#{prefix}/Notice.txt", "#{libexec}/Notice.txt"
    mv "#{prefix}/README.txt", "#{libexec}/README.txt"
```

And the output says the following:
```bash
/usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/2.6.3_2/lib/ruby/2.6.0/fileutils.rb:529:in `rename'
Errno::ENOENT: No such file or directory @ rb_file_s_rename - (/usr/local/Cellar/foo@1.0/1.0/License.html, /usr/local/Cellar/foo@1.0/1.0/libexec/License.html)
```

It looks like I'm trying to access these folders a bit prematurely.

Any pointers will be much appreciated.

**Reply by jonchang** [reply]
Probably `libexec` doesn't exist yet. So just use `libexec.mkdir`.

In general you can find a lot of similar solutions from searching existing formula files like `fgrep -lR .mkdir $(brew --repo homebrew/core)`

**Reply by gromgit** [reply]
Since you haven't shown us your _current_ formula, I can only guess that you still have:
```
    prefix.install Dir["*"]
```
at the beginning of your `install` block.

If you've not installed anything to `#{libexec}` before those lines you posted, and your source hierarchy doesn't itself contain a `libexec` subdirectory, then `#{libexec}` wouldn't exist for you to `mv` to. You're probably better off with:
```
    %w[License.html Notice.txt README.txt].each do |f|
      libexec.install "#{prefix}/#{f}"
    end
```

**Reply by cesarcoatl** [reply]
Updated version:

```ruby
class FooAT10 < Formula
  desc "Awesome description"
  homepage "https://foo.com/"
  url "https://files.foo.com/release/foo/1.0/foo-1.0.zip", using: :curl, referer: "https://foo.com/"
  sha256 "7122870ff380ac0ffc22a81fd13ff7bc5de1a0736ac39f7bed9456cfec4b799a"
  license :cannot_represent

  def install
    libexec.install Dir["*"]
    # Make files executable and add symlinks
    %w[foo.sh bar.sh ibaz].each do |cmd|
      chmod "u=wrx,go=rx", "#{libexec}/#{cmd}"
      bin.install_symlink "#{libexec}/#{cmd}" => "#{cmd.gsub(".sh", "")}-#{version}"
    end
    # Update plist file
    inreplace "#{libexec}/com.foo.plist" do |s|
      s.gsub! "/usr/local/bin/foo", "/usr/local/bin/foo-#{version}"
    end
  end

  test do
    assert_match "Foo", shell_output("#{bin}/foo-#{version} status")
  end
end
```

Folder contents:
```bash
$ pwd
/usr/local/Cellar/foo@1.0/1.0
$ ls -al
total 1016
drwxr-xr-x  10 thecesrom  admin     320 Jan 15 21:13 .
drwxr-xr-x   4 thecesrom  admin     128 Jan 15 21:13 ..
-rw-r--r--@  1 thecesrom  admin    6148 Jan 15 21:13 .DS_Store
drwxr-xr-x   3 thecesrom  admin      96 Jan 15 21:12 .brew
-rw-r--r--   1 thecesrom  admin     731 Jan 15 21:12 INSTALL_RECEIPT.json
-rw-r--r--   1 thecesrom  admin   33826 Nov 12 11:26 License.html
-rw-r--r--   1 thecesrom  admin  463145 Nov 12 11:25 Notice.txt
-rw-r--r--   1 thecesrom  admin    2060 Nov 12 11:25 README.txt
drwxr-xr-x   5 thecesrom  admin     160 Jan 15 21:12 bin
drwxr-xr-x  13 thecesrom  admin     416 Jan 15 21:12 libexec
```

**Reply by gromgit** [reply]
> ```ruby
>       s.gsub! "/usr/local/bin/foo", "/usr/local/bin/foo-#{version}"
> ```

You should use `#{bin}` here, instead of hardcoding `/usr/local/bin`.

Also, the three doc files should _not_ appear in `#{prefix}`, since nothing in your formula puts them there. Are you sure you reinstalled your formula without errors?

**Reply by cesarcoatl** [reply]
Thanks, @gromgit.

I have uninstalled and reinstalled, without errors running `brew brew install --debug --verbose foo@1.0` based on the latest Formula I posted.

An regarding those three files that remain at the root folder, that's also something that I've seen with the formula for `jython@2.7.1` I created with `brew extract --force --version=2.7.1 jython coatl-dev/coatl-dev` ([here](https://github.com/coatl-dev/homebrew-coatl-dev/blob/coatl/Formula/jython%402.7.1.rb)).

```bash
$ pwd
/usr/local/Cellar/jython@2.7.1/2.7.1
$ ls -al
total 256
drwxr-xr-x  10 thecesrom  admin     320 Jan 15 21:44 .
drwxr-xr-x   4 thecesrom  admin     128 Jan 15 21:44 ..
-rw-r--r--@  1 thecesrom  admin    6148 Jan 15 21:44 .DS_Store
drwxr-xr-x   3 thecesrom  admin      96 Jan  8 14:50 .brew
-rw-r--r--   1 thecesrom  admin     727 Jan  8 14:50 INSTALL_RECEIPT.json
-rw-r--r--   1 thecesrom  admin    9805 Jul 31  2014 LICENSE.txt
-rw-r--r--   1 thecesrom  admin  100252 Jun 17  2017 NEWS
-rw-r--r--   1 thecesrom  admin    2021 Jun 30  2017 README.txt
drwxr-xr-x   3 thecesrom  admin      96 Jan  8 14:50 bin
drwxr-xr-x  14 thecesrom  admin     448 Jan 15 17:55 libexec
```

**Reply by carlocab** [reply]
`brew install` automatically installs some files into `prefix` even if the `install` block does not tell them to. See, for example: https://github.com/Homebrew/brew/blob/master/Library/Homebrew/metafiles.rb

**Reply by cesarcoatl** [reply]
Finally, I added this line just before the `end` of the `install`.

```ruby
    # Link plist
    prefix.install_symlink "#{libexec}/com.foo.plist" => "#{plist_name}.plist"
```

And get the following when I install the formula:

```bash
$ brew install coatl-dev/coatl-dev/foo@1.0
==> Installing foo@1.0 from coatl-dev/coatl-dev
==> Downloading https://files.foo.com/release/foo/1.0/foo-1.0.zip
Already downloaded: /Users/thecesrom/Library/Caches/Homebrew/downloads/7e797d72183802dc02336411568b0a7cded5dd1b3c6ec49e1793ac8baaffc1a8--foo-1.0.zip
==> Caveats
To have launchd start coatl-dev/coatl-dev/foo@1.0 now and restart at login:
  brew services start coatl-dev/coatl-dev/foo@1.0
==> Summary
🍺  /usr/local/Cellar/foo@1.0/1.0: 1,323 files, 926.7MB, built in 39 seconds
```

I can live with those three files that remain there for some unknown reason.

Thanks @carlocab, @jonchang, and @gromgit for your input.

**Reply by cesarcoatl** [reply]
Latest finding.

By placing "move" instructions within `post_install` has solved the "misplaced" files problem:

```ruby
  def post_install
    %w[License.html Notice.txt README.txt].each do |f|
      libexec.install "#{prefix}/#{f}"
    end
  end
```
